### PR TITLE
Python based device scanning interfacing with native C++ code on linux.

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Native C++ device scanning is not implemented nor accessible from python, which does not allow us to verify/script the parsing of discovered devices.

Current discovery is done natively by Android and iOS platforms and a pure python version. However connectivity itself is done by C++ native code which would only be exercised on connect and never on discovery.

 #### Summary of Changes
 This is a merge of several branches developed while the bluez code split was under review. It contains:
 
 - Moved BlueZ main loop to a separate class (rather than inline method) so that it is accessible by other implementations
 - Implemented a 'CHIPDeviceScanner' that can be used both externally and is used internally by the 'connect via BLE' implementation in linux BLEManager
 - Added native handle classes for python for managing native interface calls and doing 'init' for platform-specific stack bits (e.g. memory and device layer initialization required by BLE)
 - Added bindings to the device scanner from python
 - Added more type safety to BlueZ code (less void *, replaced with actual types and avoided casting)
 
 Example usage:
 
 ```
 In [1]: import chip.ble

In [2]: chip.ble.DiscoverAsync(2000, lambda *x: print('Discovered: %r' % (x,)), lambda: print('Done'))
CHIP:DL: Platform main loop started.
CHIP:DL: TRACE: Bluez mainloop starting Thread
CHIP:BLE: BLE scanning through known devices.
CHIP:BLE: BLE initiating scan.

In [3]: CHIP:BLE: Device 38:CB:9F:C4:D5:7A does not look like a CHIP device.
CHIP:BLE: Device 79:1C:AB:8D:B4:1E does not look like a CHIP device.
CHIP:BLE: Device 6E:91:AB:C1:8B:75 does not look like a CHIP device.
CHIP:BLE: Device 53:DA:56:14:DF:2E does not look like a CHIP device.
Discovered: (b'98:F4:AB:6B:1A:FE', 3840, 65279, 9050)
Done
In [3]:
 ```

### CODE REVIEW

File: src/platform/Linux/bluez/MainLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        if (mBluezMainLoop != nullptr)
+        {
+            g_free(mBluezMainLoop);
+            mBluezMainLoop = nullptr;
+        }
+    }
```
Comment (msandstedt) [inline src/platform/Linux/bluez/MainLoop.cpp:85]: This cleanup logic looks suspicious.  err will always equal CHIP_NO_ERROR, so this block can never execute.  But if it was to execute, shouldn't the pthread also be cleaned up?

**Review by woody-apple — APPROVED** [review]

File: src/platform/Linux/bluez/ChipDeviceScanner.cpp [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+#include <pthread.h>
+#include <support/ReturnMacros.h>
+#include <support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace DeviceLayer {
+namespace Internal {
+namespace {
+
+struct GObjectUnref
```
Comment (mspang) [inline src/platform/Linux/bluez/ChipDeviceScanner.cpp:36]: Can we use this in other cases where we're manually calling g_object_unref?

File: src/platform/Linux/bluez/ChipDeviceScanner.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    virtual void OnScanComplete() = 0;
+};
+
+/// Allows scanning for CHIP devices
+///
+/// Will perform scan operations and call back whenever a device is discovered.
+class ChipDeviceScanner
+{
+public:
+    using Ptr = std::unique_ptr<ChipDeviceScanner>;
```
Comment (mspang) [inline src/platform/Linux/bluez/ChipDeviceScanner.h:55]: This is less readable than just spelling out `std::unique_ptr<ChipDeviceScanner>` everywhere. Please do that.

File: src/platform/Linux/bluez/MainLoop.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    mBluezMainLoop = g_main_loop_new(nullptr, FALSE);
+    VerifyOrExit(mBluezMainLoop != nullptr, ChipLogError(DeviceLayer, "FAIL: memory alloc in %s", __func__));
+
+    pthreadErr = pthread_create(&mThread, nullptr, &MainLoop::Thread, reinterpret_cast<void *>(this));
+    tmpErrno   = errno;
+    VerifyOrExit(pthreadErr == 0, ChipLogError(DeviceLayer, "FAIL: pthread_create (%s) in %s", strerror(tmpErrno), __func__));
+
+    while (!g_main_loop_is_running(mBluezMainLoop))
+    {
+        pthread_yield();
```
Comment (mspang) [inline src/platform/Linux/bluez/MainLoop.cpp:73]: Why is this needed?

File: src/platform/Linux/bluez/Helper.cpp [hunk 5]
```diff
@@ -1436,60 +1338,30 @@ static void * BluezMainLoop(void * apClosure)
...
-    g_main_loop_run(sBluezMainLoop);
-    ChipLogDetail(DeviceLayer, "TRACE: Bluez mainloop stopping %s", __func__);
+    g_signal_connect(manager, "object-added", G_CALLBACK(BluezSignalOnObjectAdded), endpoint);
+    g_signal_connect(manager, "object-removed", G_CALLBACK(BluezSignalOnObjectRemoved), endpoint);
+    g_signal_connect(manager, "interface-proxy-properties-changed", G_CALLBACK(BluezSignalInterfacePropertiesChanged), endpoint);
 
-    BluezObjectsCleanup(endpoint);
+    if (!MainLoop::Instance().SetCleanupFunction(BluezObjectsCleanup, endpoint))
+    {
+        ChipLogError(DeviceLayer, "Failed to schedule cleanup function");
```
Comment (mspang) [inline src/platform/Linux/bluez/Helper.cpp:1347]: Why isn't this an assert?

Printing a message when things go badly off the rails isn't going to catch issues quickly and reliably (e.g., cause our tests to fail).

File: src/platform/Linux/bluez/MainLoop.cpp (hunk 4 above)
Comment (msandstedt) [reply src/platform/Linux/bluez/MainLoop.cpp:73]: It appears it is needed because later on we have these checks:

```
bool MainLoop::RunOnBluezThread(GSourceFunc callback, void * argument)
{
    GMainContext * context = nullptr;
    const char * msg       = nullptr;

    VerifyOrExit(mBluezMainLoop != nullptr, msg = "FAIL: NULL sBluezMainLoop");
    VerifyOrExit(g_main_loop_is_running(mBluezMainLoop), msg = "FAIL: sBluezMainLoop not running");
```

Having said that, do we need these checks?  I'm not aware of any problems that would be created by enqueuing tasks on the event loop before g_main_loop_run has been called.

Is the worry that Bluez tasks could be erroneously scheduled after the loop quits and we need to catch this?  Personally, I would advocate that the thread creation and g_main_loop_run

File: src/platform/Linux/bluez/ChipDeviceScanner.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    void ReportDevice(BluezDevice1 * device);
+
+    GDBusObjectManager * mManager         = nullptr;
+    BluezAdapter1 * mAdapter              = nullptr;
+    GCancellable * mCancellable           = nullptr;
+    ChipDeviceScannerDelegate * mDelegate = nullptr;
+    gulong mObjectAddedSignal             = 0;
+    gulong mInterfaceChangedSignal        = 0;
+    volatile bool mIsScanning             = false;
+    volatile bool mIsStopping             = false;
```
Comment (msandstedt) [inline src/platform/Linux/bluez/ChipDeviceScanner.h:98]: What is the thread safety story here for the bluez loop?  Making these flags volatile isn't really the same as proper synchronization.

Also, `MainLoopStopScan` is calling this directly: `delegate->OnScanComplete();`.  It seems problematic for calling code to have callbacks executed in the bluez thread context.

Comment (mspang) [reply src/platform/Linux/bluez/ChipDeviceScanner.h:98]: +1

volatile has useful threading semantics in Java but not in C or C++. The equivalent in C++ is std::atomic.

File: src/platform/Linux/bluez/ChipDeviceScanner.cpp (hunk 2 above)
Comment (andy31415) [reply src/platform/Linux/bluez/ChipDeviceScanner.cpp:36]: Eventually I assume yes. I think it should be pulled out more globally as we change this code more.

I wonder though how much we will still touch it: I wanted to have scanning and this PR has it. The only additonal thing we may want is BLE connection by address instead of discriminator.

File: src/platform/Linux/bluez/ChipDeviceScanner.h (hunk 3 above)
Comment (andy31415) [reply src/platform/Linux/bluez/ChipDeviceScanner.h]: Done.

File: src/platform/Linux/bluez/ChipDeviceScanner.h (hunk 6 above)
Comment (andy31415) [reply src/platform/Linux/bluez/ChipDeviceScanner.h]: Updated to use 'ScheduleAndWait' for things that I needed executed in a separate thread and complete (instead of looping on a variable and yielding)

File: src/platform/Linux/bluez/MainLoop.cpp (hunk 4 above)
Comment (andy31415) [reply src/platform/Linux/bluez/MainLoop.cpp]: I could not find a clear way to resolve this. I need 'gmainloop' to actually run otherwise other scheduling will fail, yet I don't seem to have a clear callback of when this completes.

File: src/platform/Linux/bluez/MainLoop.cpp (hunk 1 above)
Comment (andy31415) [reply src/platform/Linux/bluez/MainLoop.cpp:132]: Fixed - I did not notice no error was ever set; thansk for catching.

File: src/controller/python/chip/ble/LinuxImpl.cpp [hunk 7]
```diff
@@ -11,41 +12,126 @@ using namespace chip::DeviceLayer::Internal;
...
+        {
+            mCompleteCallback(mContext);
+        }
+
+        delete this;
+    }
+
+private:
+    ChipDeviceScanner::Ptr mScanner;
+    PyObject * mContext;
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/controller/python/chip/ble/LinuxImpl.cpp]: Updated. Made me not be able to set it to nullptr in destructor (I was trying too set everything to null to catch any odd reference errors while debugging what turned out to be python void* cast issues).

File: src/controller/python/chip/ble/LinuxImpl.cpp [hunk 8]
```diff
@@ -11,41 +12,126 @@ using namespace chip::DeviceLayer::Internal;
...
+                                           uint16_t productId);
+    using ScanCompleteCallback  = void (*)(PyObject * context);
+
+    ScannerDelegateImpl(PyObject * context, DeviceScannedCallback scanCallback, ScanCompleteCallback completeCallback) :
+        mContext(context), mScanCallback(scanCallback), mCompleteCallback(completeCallback)
+    {}
+
+    ~ScannerDelegateImpl()
+    {
+        mContext          = nullptr;
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/controller/python/chip/ble/LinuxImpl.cpp]: Debugging invalid memory references. Turned out to be python c_void_p behavind oddly (it casts to 32-bit if it can, but when passing back it stays 32-bit int and does a sign extension or invalid high order bits when sent as a function argument ... very frustrating)

File: src/controller/python/chip/ble/scan_devices.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  closure.ScanCompleted()
+
+
+def DiscoverAsync(timeoutMs: int, scanCallback, doneCallback, adapter=None):
+  """Initiate a BLE discovery of devices with the given timeout.
+
+  NOTE: devices are not guaranteed to be unique. New entries are returned
+  as soon as the underlying BLE manager detects changes.
+
+  :param timeoutMs    - scan will complete after this time
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/controller/python/chip/ble/scan_devices.py]: Done.

File: src/platform/Linux/bluez/MainLoop.cpp (hunk 4 above)
Comment (andy31415) [reply src/platform/Linux/bluez/MainLoop.cpp]: Updated to schedule things on the main thread and wait on a semaphore.

File: src/controller/python/chip/ble/types.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+from ctypes import CFUNCTYPE, py_object, c_char_p, c_uint16
+
+DeviceScannedCallback_t = CFUNCTYPE(None, py_object, c_char_p, c_uint16,
```
[the remark this answers is not in the document]
Comment (mspang) [reply src/controller/python/chip/ble/types.py]: IMO it is not needed especially as it sounds like a C style but doesn't match the C++ typename.

File: src/controller/python/chip/ble/LinuxImpl.cpp [hunk 11]
```diff
@@ -11,41 +12,125 @@ using namespace chip::DeviceLayer::Internal;
...
+        if (mCompleteCallback)
+        {
+            mCompleteCallback(mContext);
+        }
+
+        delete this;
+    }
+
+private:
+    std::unique_ptr<ChipDeviceScanner> mScanner;
```
Comment (mspang) [inline src/controller/python/chip/ble/LinuxImpl.cpp:107]: nit: const

File: src/controller/python/chip/ble/LinuxImpl.cpp [hunk 12]
```diff
@@ -11,41 +12,125 @@ using namespace chip::DeviceLayer::Internal;
...
+    PyObject * const mContext;
+    DeviceScannedCallback mScanCallback;
```
Comment (mspang) [inline src/controller/python/chip/ble/LinuxImpl.cpp:109]: nit: const

File: src/controller/python/chip/ble/LinuxImpl.cpp [hunk 13]
```diff
@@ -11,41 +12,125 @@ using namespace chip::DeviceLayer::Internal;
...
+    ScanCompleteCallback mCompleteCallback;
```
Comment (mspang) [inline src/controller/python/chip/ble/LinuxImpl.cpp:110]: nit: const

File: src/controller/python/chip/ble/library_handle.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  handle = chip.native.GetLibraryHandle()
+
+  # Uses one of the type decorators as an indicator for everything being
+  # initialized. Native methods default to c_int return types
+  if handle.pychip_ble_adapter_list_new.restype != VoidPointer:
+    setter = chip.native.NativeLibraryHandleMethodArguments(handle)
+
+    setter.Set('pychip_ble_adapter_list_new', VoidPointer, [])
+    setter.Set('pychip_ble_adapter_list_next', c_bool, [VoidPointer])
+    setter.Set('pychip_ble_adapter_list_get_index', c_uint, [VoidPointer])
```
Comment (mspang) [inline src/controller/python/chip/ble/library_handle.py:47]: Can we do unsigned int -> uint32_t here and for pychip_ble_start_scanning?

[background: the CHIP and Google style guides recommends size types. Actually Google's style prohibits unsized types like `unsigned` altogether, except for `int`]

File: src/platform/Linux/bluez/MainLoop.cpp [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    Semaphore semaphore;
+
+    GMainContext * context = g_main_loop_get_context(mBluezMainLoop);
+    if (context == nullptr)
+    {
+        ChipLogError(DeviceLayer, "Unexpected null context on the main loop");
+        // bluez loop not cleaned as running thread is already started.
+        return CHIP_ERROR_INTERNAL;
```
Comment (mspang) [inline src/platform/Linux/bluez/MainLoop.cpp:141]: free the main loop?

**Review by mspang — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/platform/Linux/bluez/MainLoop.cpp (hunk 15 above)
Comment (andy31415) [reply src/platform/Linux/bluez/MainLoop.cpp:141]: I cannot free it because the underlying thread that will do 'run main loop' on it will be active.

We could add more logic to fix it up (maybe by schedule-and-wait on something that kills the main thread) however this seems impossible to really happen at runtime: we just created the object and it has to have a context. 

If we need to update it, I would prefer to do it separately or 'VerifyOrDie' instead to not worry about the case.

Comment (andy31415) [reply src/platform/Linux/bluez/MainLoop.cpp]: Updated to VerifyOrDie

File: src/controller/python/chip/ble/LinuxImpl.cpp (hunk 11 above)
Comment (andy31415) [reply src/controller/python/chip/ble/LinuxImpl.cpp]: Cannot do this because scanner is moved post object creation and I have a SetScanner method. 

This is a side-effect of creation order: delegate needs to access the scanner, but scanner needs a delegate.

File: src/platform/Linux/bluez/Helper.cpp (hunk 5 above)
Comment (andy31415) [reply src/platform/Linux/bluez/Helper.cpp:1347]: It may require additional passes: as it stands, we do not seem to ever stop the main loop, so I kept the cleanup function because cleanup existed in previous code, however this really does not get called in the current code structure.
